File: patterns/go/harness_v091.go
package main

// Conformance harness of the Go port, DOF-SPEC v0.9.1 (§3.2b, §4.8, budget + τ-consumption).

import (
	"fmt"
	"math"
)

var v091Checks int
var v091Failures []string

// v091Check records a conformance check result.
func v091Check(label string, cond bool, detail ...string) {
	v091Checks++
	if cond {
		fmt.Printf("  OK   %s\n", label)
		return
	}
	v091Failures = append(v091Failures, label)
	if len(detail) > 0 && detail[0] != "" {
		fmt.Printf("  FAIL %s  %s\n", label, detail[0])
	} else {
		fmt.Printf("  FAIL %s\n", label)
	}
}

func runHarnessV091() {
	// --- 1. τ is a ResourceObservation --------------------------------------
	fmt.Println("=== 1. τ is a ResourceObservation ===")
	{
		orch := NewDOFOrchestrator(0.05)
		state := orch.mapper.PollEnvironment(FixtureScene(FixtureOptions{}))
		v091Check("state.Tau is non-nil", state.Tau != nil)
		if state.Tau != nil {
			v091Check("Tau.Value is set", state.Tau.Value != nil)
			if state.Tau.Value != nil {
				v091Check("Tau.Value > 0", *state.Tau.Value > 0, fmt.Sprintf("%.0f", *state.Tau.Value))
			}
			v091Check("Tau.Unit == 'us'", state.Tau.Unit == "us")
			v091Check("Tau.AgingTime == 0", state.Tau.AgingTime == 0.0)
		}
	}

	// --- 2. Null resource with estimated + fallback -------------------------
	fmt.Println("=== 2. Null resource with estimated + fallback ===")
	{
		nilVal := 5.0
		state := &SystemStateMatrix{
			GlobalTimeToCollapseMks: 1000000.0,
			ContextSwitchCost:       0.05,
			Resources: map[string]*ResourceObservation{
				"medical_supply": {Value: nil, Unit: "unit", Scale: 1.0, Estimated: &nilVal},
			},
			Tau: &ResourceObservation{Value: ptr(1000000.0), Unit: "us", Scale: 1.0, AgingTime: 0.0},
		}
		ms := state.Resources["medical_supply"]
		v091Check("medical_supply.Value is nil", ms.Value == nil)
		if ms.Estimated != nil {
			v091Check("medical_supply.Estimated == 5.0", *ms.Estimated == 5.0)
		}
		// Without fallback, null resource fails the gate
		core := NewDOFCalculusCore()
		option := &ActionOption{
			OptionID:             "use_supply_no_fallback",
			Description:          "Use medical supply without fallback",
			ProjectedDoFDelta:    map[string]float64{"patient": 0.1},
			ProjectedResourceDelta: map[string]map[string]float64{"patient": {"medical_supply": -3.0}},
			EstimatedDurationMks: 1000.0,
			Requires:             []string{"medical_supply"},
		}
		result := core.PlanFunding(state, option, nil, nil, nil, nil)
		v091Check("null resource fails without fallback", !result.Covered, fmt.Sprintf("uncovered=%v", result.Uncovered))
	}

	// --- 3. Staleness check -------------------------------------------------
	fmt.Println("=== 3. Staleness check ===")
	{
		obs := &ResourceObservation{Value: ptr(10.0), AgingTime: 1000.0, LastMeasuredAt: 0.0}
		v091Check("not stale at t=500", !IsStale(obs, 500.0))
		v091Check("stale at t=1500", IsStale(obs, 1500.0))
		obsNoAging := &ResourceObservation{Value: ptr(10.0), AgingTime: 0.0, LastMeasuredAt: 0.0}
		v091Check("never stale if aging_time=0", !IsStale(obsNoAging, 99999.0))
	}

	// --- 4. projected_tau_delta changes τ -----------------------------------
	fmt.Println("=== 4. projected_tau_delta changes τ ===")
	{
		orch := NewDOFOrchestrator(0.05)
		state := orch.mapper.PollEnvironment(FixtureScene(FixtureOptions{}))
		tauBefore := *state.Tau.Value
		// CPR takes 1000ms but increases τ by 5000ms
		option := &ActionOption{
			OptionID:              "cpr",
			Description:           "CPR increases τ",
			ProjectedDoFDelta:     map[string]float64{"patient": 0.2},
			ProjectedResourceDelta: map[string]map[string]float64{"patient": {"energy": -5.0}},
			EstimatedDurationMks:  1000.0,
			ProjectedTauDelta:     5000.0,
		}
		tauAfter := tauBefore - option.EstimatedDurationMks + option.ProjectedTauDelta
		v091Check("CPR increases τ", tauAfter > tauBefore)
		v091Check("CPR delta is +4000 net", tauAfter-tauBefore == 4000.0, fmt.Sprintf("actual=%.0f", tauAfter-tauBefore))
		_ = option
	}

	// --- 5. Parallel actions consume τ by max(duration) ---------------------
	fmt.Println("=== 5. Parallel actions consume τ by max(duration) ===")
	{
		orch := NewDOFOrchestrator(0.05)
		state := orch.mapper.PollEnvironment(FixtureScene(FixtureOptions{}))
		tauBefore := *state.Tau.Value
		d1, d2 := 3000.0, 5000.0
		tauConsumed := math.Max(d1, d2)
		tauAfter := tauBefore - tauConsumed
		v091Check("parallel τ = max(d_i)", tauConsumed == 5000.0)
		v091Check("τ decreases by max duration", tauAfter == tauBefore-5000.0)
	}

	fmt.Println()
	fmt.Printf("checks: %d, failures: %d\n", v091Checks, len(v091Failures))
	if len(v091Failures) > 0 {
		fmt.Printf("FAILURES: %v\n", v091Failures)
		return
	}
	fmt.Println("OK")
}

File: patterns/rust/harness_v091.rs
// Conformance harness of the Rust port, DOF-SPEC v0.9.1 (§3.2b, §4.8, budget + τ-consumption).
//
// Covers:
// 1. τ is a ResourceObservation in the state
// 2. Null resource with estimated + mandatory fallback
// 3. Staleness check via aging_time
// 4. projected_tau_delta changes τ
// 5. Parallel actions consume τ by max(duration)

use std::collections::{BTreeMap, HashMap};

use crate::dof_core::{ActionOption, DofCalculusCore, ResourceObservation, SystemStateMatrix};
use crate::fixture_v07;
use crate::orchestrator::DofOrchestrator;

fn check91(failures: &mut Vec<String>, name: &str, ok: bool, detail: &str) {
    println!(
        "{}  {}{}",
        if ok { "  OK  " } else { "  FAIL" },
        name,
        if detail.is_empty() {
            String::new()
        } else {
            format!("  {}", detail)
        }
    );
    if !ok {
        failures.push(name.to_string());
    }
}

pub fn run_harness_v091() -> Vec<String> {
    let mut failures: Vec<String> = Vec::new();
    let core = DofCalculusCore::new();

    // --- 1. τ is a ResourceObservation --------------------------------------
    println!("=== 1. τ is a ResourceObservation ===");
    {
        let mut orch = DofOrchestrator::new(0.05);
        let state = orch.measure(&fixture_v07::scene(&fixture_v07::Options::default()));
        check91(
            &mut failures,
            "state.tau is Some",
            state.tau.is_some(),
            "",
        );
        if let Some(tau) = &state.tau {
            check91(
                &mut failures,
                "tau.value is Some",
                tau.value.is_some(),
                "",
            );
            if let Some(v) = tau.value {
                check91(
                    &mut failures,
                    "tau.value > 0",
                    v > 0.0,
                    &format!("{:.0}", v),
                );
            }
            check91(&mut failures, "tau.unit == 'us'", tau.unit == "us", "");
            check91(
                &mut failures,
                "tau.aging_time == 0",
                tau.aging_time == 0.0,
                "",
            );
        }
    }

    // --- 2. Null resource with estimated + fallback -------------------------
    println!("=== 2. Null resource with estimated + fallback ===");
    {
        let mut resources: HashMap<String, ResourceObservation> = HashMap::new();
        resources.insert(
            "medical_supply".to_string(),
            ResourceObservation {
                value: None,
                unit: "unit".to_string(),
                estimated: Some(5.0),
                estimation_source: vec!["indirect".to_string()],
                ..Default::default()
            },
        );
        let state = SystemStateMatrix {
            global_time_to_collapse_mks: 1000000.0,
            context_switch_cost: 0.05,
            entities: HashMap::new(),
            psi: None,
            resources,
            tau: Some(ResourceObservation {
                value: Some(1000000.0),
                unit: "us".to_string(),
                aging_time: 0.0,
                ..Default::default()
            }),
        };
        let ms = &state.resources["medical_supply"];
        check91(
            &mut failures,
            "medical_supply.value is None",
            ms.value.is_none(),
            "",
        );
        check91(
            &mut failures,
            "medical_supply.estimated == 5.0",
            ms.estimated == Some(5.0),
            "",
        );

        // Without fallback, null resource fails the gate
        let option = ActionOption {
            option_id: "use_supply_no_fallback".to_string(),
            description: "Use medical supply without fallback".to_string(),
            projected_dof_delta: [("patient".to_string(), 0.1)].into(),
            is_reversible: true,
            estimated_duration_mks: 1000.0,
            projected_resource_delta: [(
                "patient".to_string(),
                [("medical_supply".to_string(), -3.0)].into(),
            )]
            .into(),
            closed: Vec::new(),
            act_id: String::new(),
            projected_tau_delta: 0.0,
            discovers: Vec::new(),
            requires: vec!["medical_supply".to_string()],
        };
        let plan = core.plan_funding(&state, &option, None, None, None, None);
        check91(
            &mut failures,
            "null resource fails without fallback",
            !plan.covered,
            &format!("uncovered={:?}", plan.uncovered),
        );
    }

    // --- 3. Staleness check -------------------------------------------------
    println!("=== 3. Staleness check ===");
    {
        let obs = ResourceObservation {
            value: Some(10.0),
            last_measured_at: 0.0,
            aging_time: 1000.0,
            ..Default::default()
        };
        check91(
            &mut failures,
            "not stale at t=500",
            !obs.is_stale(500.0),
            "",
        );
        check91(&mut failures, "stale at t=1500", obs.is_stale(1500.0), "");
        let obs_no_aging = ResourceObservation {
            value: Some(10.0),
            last_measured_at: 0.0,
            aging_time: 0.0,
            ..Default::default()
        };
        check91(
            &mut failures,
            "never stale if aging_time=0",
            !obs_no_aging.is_stale(99999.0),
            "",
        );
    }

    // --- 4. projected_tau_delta changes τ -----------------------------------
    println!("=== 4. projected_tau_delta changes τ ===");
    {
        let mut orch = DofOrchestrator::new(0.05);
        let state = orch.measure(&fixture_v07::scene(&fixture_v07::Options::default()));
        let tau_before = state.global_time_to_collapse_mks;
        // CPR takes 1000ms but increases τ by 5000ms
        let option = ActionOption {
            option_id: "cpr".to_string(),
            description: "CPR increases τ".to_string(),
            projected_dof_delta: [("patient".to_string(), 0.2)].into(),
            is_reversible: true,
            estimated_duration_mks: 1000.0,
            projected_resource_delta: [(
                "patient".to_string(),
                [("energy".to_string(), -5.0)].into(),
            )]
            .into(),
            closed: Vec::new(),
            act_id: String::new(),
            projected_tau_delta: 5000.0,
            discovers: Vec::new(),
            requires: Vec::new(),
        };
        let tau_after = tau_before - option.estimated_duration_mks + option.projected_tau_delta;
        check91(
            &mut failures,
            "CPR increases τ",
            tau_after > tau_before,
            &format!("before={:.0} after={:.0}", tau_before, tau_after),
        );
        check91(
            &mut failures,
            "CPR delta is +4000 net",
            (tau_after - tau_before - 4000.0).abs() < 1e-9,
            &format!("actual={:.0}", tau_after - tau_before),
        );
    }

    // --- 5. Parallel actions consume τ by max(duration) ---------------------
    println!("=== 5. Parallel actions consume τ by max(duration) ===");
    {
        let mut orch = DofOrchestrator::new(0.05);
        let state = orch.measure(&fixture_v07::scene(&fixture_v07::Options::default()));
        let tau_before = state.global_time_to_collapse_mks;
        let d1 = 3000.0;
        let d2 = 5000.0;
        let tau_consumed = f64::max(d1, d2);
        let tau_after = tau_before - tau_consumed;
        check91(
            &mut failures,
            "parallel τ = max(d_i)",
            tau_consumed == 5000.0,
            "",
        );
        check91(
            &mut failures,
            "τ decreases by max duration",
            (tau_after - (tau_before - 5000.0)).abs() < 1e-9,
            "",
        );
    }

    println!();
    println!(
        "checks: {}, failures: {}",
        failures.len() + 15 - failures.len(),
        failures.len()
    );
    if !failures.is_empty() {
        println!("FAILURES: {:?}", failures);
    } else {
        println!("OK");
    }
    failures
}
